File: python/y2021/d01/aoc_2021_01.py
import sys
from common import timer, Puzzle


class Day01(Puzzle, year=2021, day=1):
    @staticmethod
    def clean_input(data):
        return [int(d) for d in data]

    def solution(self, window):
        return len(
            [(a, b) for a, b in zip(self.data, self.data[window:]) if b > a]
        )

    @timer(part=1)
    def solve_part_1(self):
        """ Solution for part 1 """
        return self.solution(window=1)

    @timer(part=2)
    def solve_part_2(self):
        """ Solution for part 2 """
        return self.solution(window=3)

    def solve_all(self):
        part_1 = self.solve_part_1()
        part_2 = self.solve_part_2()

        return part_1, part_2


if __name__ == "__main__":
    filename = "input.txt"
    if len(sys.argv) > 1:
        filename = sys.argv[1]
    Day01(filename=filename)

File: C++/2021/aoc_2021_01.cpp
#include "aoc_io.hpp"
#include "aoc_timer.hpp"
#include <string>
#include <vector>

namespace aoc_2021_01 {

int solve_part_1(const std::vector<int> &data) {
    int sum = 0;
    for (std::vector<int>::const_iterator it1 = data.begin(), it2 = it1 + 1; it2 != data.end(); it1++, it2++) {
        if (*it1 < *it2) {
            sum++;
        }
    }

    return sum;
}

int solve_part_2(const std::vector<int> &data) {
    int sum = 0;
    for (std::vector<int>::const_iterator it1 = data.begin(), it2 = it1 + 3; it2 != data.end(); it1++, it2++) {
        if (*it1 < *it2) {
            sum++;
        }
    }

    return sum;
}

int solve_all(const std::vector<int> &data) {
    aoc::timer(1, solve_part_1, data);
    aoc::timer(2, solve_part_2, data);

    return 0;
}
} // namespace aoc_2021_01

int main(int argc, char **argv) {
    std::string filename;
    const int   year = 2021;
    const int   day  = 1;

    if (argc > 1) {
        if (std::string(argv[1]) == "--test") {
            filename = "test_input.txt";
        } else {
            filename = argv[1];
        }
    } else {
        filename = "input.txt";
    }

    std::vector<int> data;
    data = aoc::io::get_input_list<int>(filename, year, day);

    aoc::io::header(year, day);
    aoc::timer(aoc_2021_01::solve_all, data);

    return 0;
}
